preliminary request compressing (#1)
rojkov/envoy: docs/root/configuration/http/http_filters/compressor_filter.rst

the proxy won't know to fetch a new incoming request with compatible "*accept-encoding*"
from upstream.

When request compression is *applied*:

- *content-length* is removed from request headers.
- *content-encoding* with the compression scheme used (e.g., ``gzip``) is added to
  request headers.

> Using different decompressors for requests and responses

Reviewer: ```suggestion
Using different compressors for requests and responses
```

--------------------------------------------------------

If different compression libraries are desired for requests and responses, it is possible to install

> multiple decompressor filters enabled only for requests or responses. For instance:

Reviewer: ```suggestion
multiple compressor filters enabled only for requests or responses. For instance:
```

            memory_level: 3
            window_bits: 10
            compression_level: best_compression
            compression_strategy: default_strategy
    # This filter is only enabled for requests.
    - name: envoy.filters.http.compressor
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.filters.http.compressor.v3.Compressor
        response_direction_config:
          common_config:
            enabled:
              default_value: false
              runtime_key: response_compressor_enabled
        request_direction_config:
          common_config:
            enabled:
              default_value: true
              runtime_key: request_compressor_enabled
        compressor_library:
          name: for_request
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip
            memory_level: 9
            window_bits: 15
            compression_level: best_speed
            compression_strategy: default_strategy

.. _compressor-statistics:

Statistics
----------

Every configured Compressor filter has statistics rooted at

> <stat_prefix>.compressor.<compressor_library.name>.<compressor_library_stat_prefix>.<direction_prefix>.*

Reviewer: we should add release notes that mention that this stat tree is changing

  compressed, Counter, Number of requests compressed.
  not_compressed, Counter, Number of requests not compressed.
  total_uncompressed_bytes, Counter, The total uncompressed bytes of all the requests that were marked for compression.
  total_compressed_bytes, Counter, The total compressed bytes of all the requests that were marked for compression.

>   content_length_too_small, Counter, Number of requests that accepted gzip encoding but did not compress because the payload was too small.

Reviewer: ```suggestion
  content_length_too_small, Counter, Number of requests that accepted the compressor encoding but did not compress because the payload was too small.
```

In addition to the statics common for requests and responses there are statistics
specific to responses only:

.. csv-table::
  :header: Name, Type, Description
  :widths: 1, 1, 2

  no_accept_header, Counter, Number of requests with no accept header sent.
  header_identity, Counter, Number of requests sent with "identity" set as the *accept-encoding*.

>   header_compressor_used, Counter, Number of requests sent with "gzip" set as the *accept-encoding*.

Reviewer: ```suggestion
  header_compressor_used, Counter, Number of requests sent with filter's configured encoding set as the *accept-encoding*.
```
